src/lib/Cashlink.ts

```diff
@@ -0,0 +1,453 @@
+
+        // Only exit if the cashlink is CLAIMED and not currently funded or being funded.
+        if (this.state === CashlinkState.CLAIMED && !balance && !pendingFundingTx) return;
+
+        const knownTransactionReceipts = new Map(this._knownTransactions.map((tx) => [tx.hash, tx.blockHash!]));
+
+        const transactionHistory = await (await this._network).requestTransactionHistory(
+            this.address.toUserFriendlyAddress(),
+            knownTransactionReceipts,
+        );
+        this._knownTransactions = this._knownTransactions.concat(transactionHistory.newTransactions);
+
+        let newState: CashlinkState = this.state;
+
+        const knownFundingTx = this._knownTransactions.find(
+            (tx) => tx.recipient === this.address.toUserFriendlyAddress());
+        const knownClaimingTx = this._knownTransactions.find(
+            (tx) => tx.sender === this.address.toUserFriendlyAddress());
+
+        switch (this.state) {
+            case CashlinkState.UNKNOWN:
+                if (!pendingFundingTx && !knownFundingTx) {
+                    newState = CashlinkState.UNCHARGED;
+                    break;
+                }
+            case CashlinkState.UNCHARGED:
+                if (pendingFundingTx) {
+                    newState = CashlinkState.CHARGING;
+                }
+            case CashlinkState.CHARGING:
+                if (!balance && !pendingFundingTx) {
+                    // Handle expired/replaced funding tx
+                    newState = CashlinkState.UNCHARGED;
+                    // Not break;ing here, because we need to see if the cashlink is already CLAIMED.
+                }
+                if (knownFundingTx) {
+                    newState = CashlinkState.UNCLAIMED;
+                } else break; // If no known transactions are found, no further checks are necessary
+            case CashlinkState.UNCLAIMED:
+                if (pendingClaimingTx) {
+                    newState = CashlinkState.CLAIMING;
+                }
+            case CashlinkState.CLAIMING:
+                if (balance) {
+                    // Handle recharged/reused cashlink
+                    if (!pendingClaimingTx) newState = CashlinkState.UNCLAIMED;
+                    break; // If a balance is detected on the cashlink, it cannot be CLAIMED.
+                }
+                if (knownClaimingTx) {
+                    newState = CashlinkState.CLAIMED;
+                }
+            case CashlinkState.CLAIMED:
+                if (pendingFundingTx) newState = CashlinkState.CHARGING;
+                if (balance) newState = CashlinkState.UNCLAIMED;
+                if (pendingClaimingTx) newState = CashlinkState.CLAIMING;
+        }
+
+        if (newState !== this.state) this._updateState(newState);
+    }
+
+    public toObject(): CashlinkEntry {
+        return {
+            address: this.address.toUserFriendlyAddress(),
+            keyPair: new Uint8Array(this.keyPair.serialize()),
```

Comment:
I believe the main reason is that the tests fail because of differing types.

Reply:
And if the tests would be changed to expect a `SerialBuffer`?

Otherwise, to avoid the copy that is involved here, `new Uint8Array(serialized.buffer)` could be used.

Reply:
I checked it again and it's the difference between what the Cashlink uses internally, compared with what gets stored and returned in the IndexedDB. But I was able to remove all conversions except the one in `toObject()`.